src/Core/StreamingHandleErrorMode.h

```diff
@@ -9,7 +9,13 @@ enum class StreamingHandleErrorMode : uint8_t
 {
     DEFAULT = 0, // Ignore errors with threshold.
     STREAM, // Put errors to stream in the virtual column named ``_error.
-    /*FIXED_SYSTEM_TABLE, Put errors to in a fixed system table likely system.kafka_errors. This is not implemented now.  */
+};
+
+enum class ExtStreamingHandleErrorMode : uint8_t
+{
+    DEFAULT = 0, // Ignore errors with threshold.
+    STREAM, // Put errors to stream in the virtual column named ``_error.
+    DEAD_LETTER_QUEUE
     /*CUSTOM_SYSTEM_TABLE, Put errors to in a custom system table. This is not implemented now.  */
 };
```

Comment:
Have you thought about just extending `StreamingHandleErrorMode` and throw an error when `dead_letter_queue` is specified for an engine that doesn't support that? For me that feels like a better approach, because the two enums are really about the same setting.

Reply:
Honestly I don't think that it is a right thing to do, basically it compromises the idea of typed enums. As a practical reason - if we use extended enum everywhere we will  struggle with compiler warnings.
What we need is inheritance, but it is not possible out of the box.

Reply:
Ahh, okay. I thought that we can set the default error handling mode in the config, but after checking it I realized it is not possible. I think this should fine then.

Reply:
I couldn't stop thinking about this and asked some of the collegues. In the end I think we should unify these enums. My reasons:
1. What if somebody wants to add another way of handling errors is stream like engine? Depending on whether it is relevant to Kafka/RabbitMQ/other engines, we might want to make new enums. This doesn't feel good. If it affects all engines, then they have to extend two enums.
2. Similar to the previous one, but slightly different: if somebody wants to change what the `DEFAULT` value mean, they might miss to change the ones with the extended enum.

Therefore I think the better approach is to have a single enum and make sure that `DEAD_LETTER_QUEUE` is handled in every engines either by using the new system table or throwing an exception on table creation (similarly to other, invalid setting values).

Reply:
@antaljanosbenjamin , could you elaborate, please.

If you believe that any fancy and engine-specific techniques are to be handled individually and should not be included in a global enum, I agree.

We assume that we understand nature of streaming tables and provide some unification. Moreover, we have cross-engine dead_letter_queue. That's why I believe that introducing ExtStreamingHandleErrorMode is reasonable, risk that it would be Ext2StreamingHandleErrorMode whatever is minimal. Although it is too early to get rid of StreamingHandleErrorMode completely.

If you believe that we have to (eventually) support dead_letter_queue and switch to ExtStreamingHandleErrorMode for all streaming engine, I agree.

If you believe that we have to switch to ExtStreamingHandleErrorMode for all streaming engine even if a particular streaming engine does not support dead_letter_queue, I would NOT agree. What is the point?

As you may remember, initially dead_letter_queue was requested for kafka. I've thought that it would be nice to have it in a more generic way, while I don't have enough courage to handle engines like NATS I know nothing about.


If you do believe that it is a good idea to use ExtStreamingHandleErrorMode for all streaming engines right now despite my considerations above - let me know, I'll implement this, it's not a big deal.

Reply:
> If you believe that we have to (eventually) support dead_letter_queue and switch to ExtStreamingHandleErrorMode for all streaming engine, I agree.

Agree. We don't have to support that now for all streaming engines.

> If you believe that we have to switch to ExtStreamingHandleErrorMode for all streaming engine even if a particular streaming engine does not support dead_letter_queue, I would NOT agree. What is the point?

The point is not to split streaming engine settings. In ClickHouse I think there quite some things that are handled in a way that some engines throws if they don't support it. Not the best, but an example is StorageKafka(2) [only supports](https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/Kafka/StorageKafkaUtils.cpp#L225-L232) ordinary columns but not `DEFAULT/MATERIALIZED/EPHEMERAL` columns. It is not the best example, because it depends on our grammar, not a single enum, but it proves that you can easily specify a column/setting for an engine which it cannot handle. In my opinion having an extra check in all streaming engines that doesn't support dead letter queue is much better solution than having two enums that are almost the same. 

> As you may remember, initially dead_letter_queue was requested for kafka. I've thought that it would be nice to have it in a more generic way, while I don't have enough courage to handle engines like NATS I know nothing about.

And that is perfectly fine. I think you have already over delivered it by implementing it to RabbitMQ, I and I think the whole community appreciates it. 

> If you genuinely believe that implementing ExtStreamingHandleErrorMode across all streaming engines is a good idea despite my concerns, please let me know, and I'll take care of it - it's not a major undertaking.
Probably we can introduce error handling traits on a per-engine basis.

Just to be clear: we don't have to support dead letter queue in all streaming engines, but we should use the same enum for error handling mode across all streaming engines and just throw an exception when the engine doesn't support it.

How does this sound? If you are still not convinced that unifying these enums is better, then feel to say your concerns or point out if I haven't addressed your already mentioned concerns properly. I would like to reach a compromise here, because I truly beleive that lack of compromise means some of the parties (that can easily be me) are missing something important.